File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/MLBF5SplitsStore.swift
import Foundation
import Observation
import Supabase
import WagerproofModels
import WagerproofServices

/// Loads tonight's MLB First-Five (F5) splits slate. 1:1 port of RN
/// `hooks/useMLBF5Splits.ts`:
///   1. `mlb_games_today` for today..+2 (drop postponed)
///   2. `mlb_team_mapping` to resolve team name → abbreviation
///   3. `mv_mlb_f5_team_splits` filtered to the slate's team abbreviations
///
/// All three live on the CFB (sports-data) Supabase, read anonymously.
@Observable
@MainActor
public final class MLBF5SplitsStore {
    public enum LoadState: Equatable, Sendable {
        case idle
        case loading
        case loaded
        case failed(String)
    }

    public private(set) var games: [MLBF5Game] = []
    public private(set) var splitLookup: [String: MLBF5SplitRow] = [:]
    public private(set) var lastRefreshedAt: String?
    public private(set) var loadState: LoadState = .idle

    public var isLoading: Bool {
        if case .loading = loadState { return true }
        return false
    }
    public var errorMessage: String? {
        if case .failed(let m) = loadState { return m }
        return nil
    }

    public init() {}

    public func refresh() async {
        loadState = .loading
        let cfb = await CFBSupabase.shared.client
        let today = Self.todayET()
        let end = Self.addDays(today, 2)

        // 1. Today's slate.
        let scheduleRows: [ScheduleRow]
        do {
            scheduleRows = try await cfb
                .from("mlb_games_today")
                .select()
                .gte("official_date", value: today)
                .lte("official_date", value: end)
                .order("official_date", ascending: true)
                .order("game_time_et", ascending: true)
                .execute()
                .value
        } catch {
            loadState = .failed(error.localizedDescription)
            return
        }

        // 2. Team mapping (best-effort; the static MLBTeams map is the fallback).
        let mappingRows: [MLBTeamMapping] = (try? await cfb
            .from("mlb_team_mapping")
            .select()
            .execute()
            .value) ?? []
        var byName: [String: MLBTeamMapping] = [:]
        var byId: [Int: MLBTeamMapping] = [:]
        for m in mappingRows {
            if !m.teamName.isEmpty { byName[MLBTeams.normalize(m.teamName)] = m }
            if m.mlbApiId != 0 { byId[m.mlbApiId] = m }
        }

        let builtGames: [MLBF5Game] = scheduleRows
            .filter { !($0.isPostponed ?? false) }
            .map { row in
                let awayName = row.awayName
                let homeName = row.homeName
                let awayAbbr = MLBF5.toSplitTeamAbbr(Self.resolveTeam(awayName, teamId: row.awayId, byName: byName, byId: byId))
                let homeAbbr = MLBF5.toSplitTeamAbbr(Self.resolveTeam(homeName, teamId: row.homeId, byName: byName, byId: byId))
                return MLBF5Game(
                    gamePk: row.gamePk,
                    officialDate: row.officialDate,
                    gameTimeEt: row.gameTimeEt,
                    awayTeamName: awayName,
                    homeTeamName: homeName,
                    venueName: row.venueName,
                    awayAbbr: awayAbbr,
                    homeAbbr: homeAbbr,
                    awaySpName: row.awaySpName,
                    homeSpName: row.homeSpName,
                    awaySpHand: MLBF5.normalizePitchHand(row.awaySpHand),
                    homeSpHand: MLBF5.normalizePitchHand(row.homeSpHand),
                    totalLine: row.totalLine,
                    f5AwayMl: row.f5AwayMl,
                    f5HomeMl: row.f5HomeMl,
                    f5TotalLine: row.f5TotalLine
                )
            }

        // 3. Splits for the slate's teams.
        let abbrs = Array(Set(builtGames.flatMap { [$0.awayAbbr, $0.homeAbbr] })).sorted()
        if abbrs.isEmpty {
            games = builtGames
            splitLookup = [:]
            loadState = .loaded
            return
        }

        let splitRows: [MLBF5SplitRow] = (try? await cfb
            .from("mv_mlb_f5_team_splits")
            .select()
            .in("team_abbr", values: abbrs)
            .execute()
            .value) ?? []

        games = builtGames
        splitLookup = MLBF5.buildSplitLookup(splitRows)
        lastRefreshedAt = splitRows.compactMap { $0.lastRefreshedAt }.first
        loadState = .loaded
    }

    public func split(for game: MLBF5Game, side: String) -> MLBF5SplitRow? {
        if side == "away" {
            return MLBF5.findSplitRow(splitLookup, teamAbbr: game.awayAbbr, homeAway: "away", oppSpHand: game.homeSpHand)
        } else {
            return MLBF5.findSplitRow(splitLookup, teamAbbr: game.homeAbbr, homeAway: "home", oppSpHand: game.awaySpHand)
        }
    }

    // MARK: - Team resolution (mirrors RN `resolveTeam`)

    private static func resolveTeam(
        _ teamName: String,
        teamId: Int?,
        byName: [String: MLBTeamMapping],
        byId: [Int: MLBTeamMapping]
    ) -> String {
        let nameKey = MLBTeams.normalize(teamName)
        if let direct = byName[nameKey], !direct.team.isEmpty { return direct.team }
        if let teamId, let byTeamId = byId[teamId], !byTeamId.team.isEmpty { return byTeamId.team }
        for (key, mapping) in byName where key.contains(nameKey) || nameKey.contains(key) {
            if !mapping.team.isEmpty { return mapping.team }
        }
        if let info = MLBTeams.info(for: teamName) { return info.team }
        // Last-ditch: initials / first chars.
        let words = teamName.split(separator: " ")
        if words.count >= 2 {
            return words.suffix(3).map { String($0.prefix(1)) }.joined().uppercased()
        }
        return String(teamName.prefix(3)).uppercased()
    }

    // MARK: - Date helpers (ET, mirrors RN getTodayET / addDaysYmd)

    static func todayET() -> String {
        let fmt = DateFormatter()
        fmt.locale = Locale(identifier: "en_CA") // yyyy-MM-dd
        fmt.timeZone = TimeZone(identifier: "America/New_York")
        fmt.dateFormat = "yyyy-MM-dd"
        return fmt.string(from: Date())
    }

    static func addDays(_ baseYmd: String, _ days: Int) -> String {
        let parser = DateFormatter()
        parser.locale = Locale(identifier: "en_US_POSIX")
        parser.timeZone = TimeZone(identifier: "America/New_York")
        parser.dateFormat = "yyyy-MM-dd"
        guard let base = parser.date(from: baseYmd),
              let shifted = Calendar.current.date(byAdding: .day, value: days, to: base) else {
            return baseYmd
        }
        return parser.string(from: shifted)
    }
}

// MARK: - Flexible schedule row decode

/// `mlb_games_today` is wide and column names drift across migrations, so we
/// read each field from the first present of several candidate keys — mirrors
/// the `||` fallbacks in the RN hook's `fetchF5Games`.
private struct ScheduleRow: Decodable {
    let gamePk: Int
    let officialDate: String
    let gameTimeEt: String?
    let awayName: String
    let homeName: String
    let venueName: String?
    let awayId: Int?
    let homeId: Int?
    let awaySpName: String?
    let homeSpName: String?
    let awaySpHand: String?
    let homeSpHand: String?
    let totalLine: Double?
    let f5AwayMl: Double?
    let f5HomeMl: Double?
    let f5TotalLine: Double?
    let isPostponed: Bool?

    struct AnyKey: CodingKey {
        var stringValue: String
        var intValue: Int? { nil }
        init?(stringValue: String) { self.stringValue = stringValue }
        init?(intValue: Int) { nil }
        init(_ s: String) { self.stringValue = s }
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: AnyKey.self)
        gamePk = c.firstInt(["game_pk", "gamePk"]) ?? 0
        officialDate = c.firstString(["official_date", "officialDate", "game_date_et"]) ?? ""
        gameTimeEt = c.firstString(["game_time_et", "gameTimeEt"])
        awayName = c.firstString(["away_team_name", "away_team", "away_team_full_name"]) ?? "Away"
        homeName = c.firstString(["home_team_name", "home_team", "home_team_full_name"]) ?? "Home"
        venueName = c.firstString(["venue_name", "venue"])
        awayId = c.firstInt(["away_team_id", "away_mlb_team_id", "away_id"])
        homeId = c.firstInt(["home_team_id", "home_mlb_team_id", "home_id"])
        awaySpName = c.firstString(["away_sp_name"])
        homeSpName = c.firstString(["home_sp_name"])
        awaySpHand = c.firstString(["away_sp_hand"])
        homeSpHand = c.firstString(["home_sp_hand"])
        totalLine = c.firstDouble(["total_line", "game_total_line"])
        f5AwayMl = c.firstDouble(["f5_away_ml"])
        f5HomeMl = c.firstDouble(["f5_home_ml"])
        f5TotalLine = c.firstDouble(["f5_total_line"])
        isPostponed = c.firstBool(["is_postponed", "postponed"])
    }
}

private extension KeyedDecodingContainer where Key == ScheduleRow.AnyKey {
    func firstString(_ keys: [String]) -> String? {
        for raw in keys {
            guard let key = ScheduleRow.AnyKey(stringValue: raw) else { continue }
            if let s = try? decode(String.self, forKey: key), !s.isEmpty { return s }
            if let i = try? decode(Int.self, forKey: key) { return String(i) }
            if let d = try? decode(Double.self, forKey: key) { return String(d) }
        }
        return nil
    }
    func firstInt(_ keys: [String]) -> Int? {
        for raw in keys {
            guard let key = ScheduleRow.AnyKey(stringValue: raw) else { continue }
            if let i = try? decode(Int.self, forKey: key) { return i == 0 ? nil : i }
            if let d = try? decode(Double.self, forKey: key) { return Int(d) == 0 ? nil : Int(d) }
            if let s = try? decode(String.self, forKey: key), let i = Int(s) { return i == 0 ? nil : i }
        }
        return nil
    }
    func firstDouble(_ keys: [String]) -> Double? {
        for raw in keys {
            guard let key = ScheduleRow.AnyKey(stringValue: raw) else { continue }
            if let d = try? decode(Double.self, forKey: key) { return d }
            if let i = try? decode(Int.self, forKey: key) { return Double(i) }
            if let s = try? decode(String.self, forKey: key), let d = Double(s) { return d }
        }
        return nil
    }
    func firstBool(_ keys: [String]) -> Bool? {
        for raw in keys {
            guard let key = ScheduleRow.AnyKey(stringValue: raw) else { continue }
            if let b = try? decode(Bool.self, forKey: key) { return b }
            if let i = try? decode(Int.self, forKey: key) { return i != 0 }
            if let s = try? decode(String.self, forKey: key) { return s == "true" || s == "t" || s == "1" }
        }
        return nil
    }
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/MLBF5SplitsStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.models.MLBF5
import com.wagerproof.core.models.MLBF5Game
import com.wagerproof.core.models.MLBF5Matchup
import com.wagerproof.core.models.MLBF5SplitRow
import com.wagerproof.core.models.MLBTeamMapping
import com.wagerproof.core.models.MLBTeams
import com.wagerproof.core.services.BuildFlags
import com.wagerproof.core.services.SupabaseClients
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Loads tonight's MLB First-Five (F5) splits slate. 1:1 port of iOS
 * `MLBF5SplitsStore.swift` (mirrors RN `hooks/useMLBF5Splits.ts`):
 *   1. `mlb_games_today` for today..+2 (drop postponed)
 *   2. `mlb_team_mapping` to resolve team name → abbreviation
 *   3. `mv_mlb_f5_team_splits` filtered to the slate's team abbreviations
 * All three live on the CFB (sports-data) Supabase, read anonymously.
 */
@Stable
class MLBF5SplitsStore {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    var games by mutableStateOf<List<MLBF5Game>>(emptyList()); private set
    var splitLookup by mutableStateOf<Map<String, MLBF5SplitRow>>(emptyMap()); private set
    var lastRefreshedAt by mutableStateOf<String?>(null); private set
    var loadState by mutableStateOf<LoadState>(LoadState.Idle); private set

    // An EMPTY slate (off-day) is cached too — otherwise every carousel swipe /
    // search keystroke would re-fire the full two-query fetch. Public so the
    // game sheets can apply the first-hydrate skeleton rule
    // (`isLoading && lastFetched == null`).
    var lastFetched by mutableStateOf<Long?>(null); private set
    private val staleWindowMs: Long = 10 * 60 * 1000

    val isLoading: Boolean get() = loadState.isLoading
    val errorMessage: String? get() = loadState.errorMessage

    fun close() = scope.cancel()

    suspend fun refreshIfStale(force: Boolean = false) {
        val last = lastFetched
        if (!force && last != null && System.currentTimeMillis() - last < staleWindowMs) return
        refresh()
    }

    suspend fun refresh() {
        loadState = LoadState.Loading
        val cfb = SupabaseClients.cfb
        val today = todayET()
        val end = addDays(today, 2)

        // 1. Today's slate.
        val scheduleRows: List<ScheduleRow>
        try {
            scheduleRows = cfb
                .from("mlb_games_today")
                .select {
                    filter {
                        gte("official_date", today)
                        lte("official_date", end)
                    }
                    order("official_date", Order.ASCENDING)
                    order("game_time_et", Order.ASCENDING)
                }
                .decodeList<JsonObject>()
                .map { ScheduleRow.from(it) }
        } catch (e: Exception) {
            loadState = LoadState.Failed(e.message ?: "Failed to load F5 splits.")
            return
        }

        // 2. Team mapping (best-effort; the static MLBTeams map is the fallback).
        val mappingRows: List<MLBTeamMapping> = runCatching {
            cfb.from("mlb_team_mapping").select().decodeList<MLBTeamMapping>()
        }.getOrDefault(emptyList())
        val byName = HashMap<String, MLBTeamMapping>()
        val byId = HashMap<Int, MLBTeamMapping>()
        for (m in mappingRows) {
            if (m.teamName.isNotEmpty()) byName[MLBTeams.normalize(m.teamName)] = m
            if (m.mlbApiId != 0) byId[m.mlbApiId] = m
        }

        val builtGames: List<MLBF5Game> = scheduleRows
            .filter { it.isPostponed != true }
            .map { row ->
                val awayName = row.awayName
                val homeName = row.homeName
                val awayAbbr = MLBF5.toSplitTeamAbbr(resolveTeam(awayName, row.awayId, byName, byId))
                val homeAbbr = MLBF5.toSplitTeamAbbr(resolveTeam(homeName, row.homeId, byName, byId))
                MLBF5Game(
                    gamePk = row.gamePk,
                    officialDate = row.officialDate,
                    gameTimeEt = row.gameTimeEt,
                    awayTeamName = awayName,
                    homeTeamName = homeName,
                    venueName = row.venueName,
                    awayAbbr = awayAbbr,
                    homeAbbr = homeAbbr,
                    awaySpName = row.awaySpName,
                    homeSpName = row.homeSpName,
                    awaySpHand = MLBF5.normalizePitchHand(row.awaySpHand),
                    homeSpHand = MLBF5.normalizePitchHand(row.homeSpHand),
                    totalLine = row.totalLine,
                    f5AwayMl = row.f5AwayMl,
                    f5HomeMl = row.f5HomeMl,
                    f5TotalLine = row.f5TotalLine,
                )
            }

        // 3. Splits for the slate's teams.
        val abbrs = builtGames.flatMap { listOf(it.awayAbbr, it.homeAbbr) }.toSortedSet().toList()
        if (abbrs.isEmpty()) {
            games = builtGames
            splitLookup = emptyMap()
            lastFetched = System.currentTimeMillis()
            loadState = LoadState.Loaded
            return
        }

        val splitRows: List<MLBF5SplitRow> = runCatching {
            cfb.from("mv_mlb_f5_team_splits")
                .select { filter { isIn("team_abbr", abbrs) } }
                .decodeList<MLBF5SplitRow>()
        }.getOrDefault(emptyList())

        games = builtGames
        splitLookup = MLBF5.buildSplitLookup(splitRows)
        lastRefreshedAt = splitRows.firstNotNullOfOrNull { it.lastRefreshedAt }
        lastFetched = System.currentTimeMillis()
        loadState = LoadState.Loaded
    }

    fun debugSet(games: List<MLBF5Game>, splits: List<MLBF5SplitRow>) {
        if (!BuildFlags.isDebugBuild) return
        this.games = games
        this.splitLookup = MLBF5.buildSplitLookup(splits)
        this.lastFetched = System.currentTimeMillis()
        this.loadState = LoadState.Loaded
    }

    /**
     * Per-game lookup for the game-sheet F5 widget. Returns null only when the
     * game isn't on the slate.
     */
    fun matchup(gamePk: Int): MLBF5Matchup? {
        val game = games.firstOrNull { it.gamePk == gamePk } ?: return null
        return MLBF5Matchup(
            game = game,
            awaySplit = split(game, "away"),
            homeSplit = split(game, "home"),
        )
    }

    fun split(game: MLBF5Game, side: String): MLBF5SplitRow? =
        if (side == "away") {
            MLBF5.findSplitRow(splitLookup, game.awayAbbr, "away", game.homeSpHand)
        } else {
            MLBF5.findSplitRow(splitLookup, game.homeAbbr, "home", game.awaySpHand)
        }

    // MARK: - Team resolution (mirrors RN `resolveTeam`)

    private fun resolveTeam(
        teamName: String,
        teamId: Int?,
        byName: Map<String, MLBTeamMapping>,
        byId: Map<Int, MLBTeamMapping>,
    ): String {
        val nameKey = MLBTeams.normalize(teamName)
        byName[nameKey]?.let { if (it.team.isNotEmpty()) return it.team }
        if (teamId != null) byId[teamId]?.let { if (it.team.isNotEmpty()) return it.team }
        for ((key, mapping) in byName) {
            if (key.contains(nameKey) || nameKey.contains(key)) {
                if (mapping.team.isNotEmpty()) return mapping.team
            }
        }
        MLBTeams.info(teamName)?.let { return it.team }
        // Last-ditch: initials / first chars.
        val words = teamName.split(" ").filter { it.isNotEmpty() }
        if (words.size >= 2) {
            return words.takeLast(3).joinToString("") { it.take(1) }.uppercase()
        }
        return teamName.take(3).uppercase()
    }

    // MARK: - Date helpers (ET, mirrors RN getTodayET / addDaysYmd)

    companion object {
        private val ET: ZoneId = ZoneId.of("America/New_York")
        private val YMD: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

        fun todayET(): String = LocalDate.now(ET).format(YMD)

        fun addDays(baseYmd: String, days: Int): String =
            runCatching { LocalDate.parse(baseYmd, YMD).plusDays(days.toLong()).format(YMD) }
                .getOrDefault(baseYmd)
    }

    // MARK: - Flexible schedule row decode

    /**
     * `mlb_games_today` is wide and column names drift across migrations, so we
     * read each field from the first present of several candidate keys — mirrors
     * the `||` fallbacks in the RN hook's `fetchF5Games`.
     */
    private data class ScheduleRow(
        val gamePk: Int,
        val officialDate: String,
        val gameTimeEt: String?,
        val awayName: String,
        val homeName: String,
        val venueName: String?,
        val awayId: Int?,
        val homeId: Int?,
        val awaySpName: String?,
        val homeSpName: String?,
        val awaySpHand: String?,
        val homeSpHand: String?,
        val totalLine: Double?,
        val f5AwayMl: Double?,
        val f5HomeMl: Double?,
        val f5TotalLine: Double?,
        val isPostponed: Boolean?,
    ) {
        companion object {
            fun from(o: JsonObject): ScheduleRow = ScheduleRow(
                gamePk = firstInt(o, listOf("game_pk", "gamePk")) ?: 0,
                officialDate = firstString(o, listOf("official_date", "officialDate", "game_date_et")) ?: "",
                gameTimeEt = firstString(o, listOf("game_time_et", "gameTimeEt")),
                awayName = firstString(o, listOf("away_team_name", "away_team", "away_team_full_name")) ?: "Away",
                homeName = firstString(o, listOf("home_team_name", "home_team", "home_team_full_name")) ?: "Home",
                venueName = firstString(o, listOf("venue_name", "venue")),
                awayId = firstInt(o, listOf("away_team_id", "away_mlb_team_id", "away_id")),
                homeId = firstInt(o, listOf("home_team_id", "home_mlb_team_id", "home_id")),
                awaySpName = firstString(o, listOf("away_sp_name")),
                homeSpName = firstString(o, listOf("home_sp_name")),
                awaySpHand = firstString(o, listOf("away_sp_hand")),
                homeSpHand = firstString(o, listOf("home_sp_hand")),
                totalLine = firstDouble(o, listOf("total_line", "game_total_line")),
                f5AwayMl = firstDouble(o, listOf("f5_away_ml")),
                f5HomeMl = firstDouble(o, listOf("f5_home_ml")),
                f5TotalLine = firstDouble(o, listOf("f5_total_line")),
                isPostponed = firstBool(o, listOf("is_postponed", "postponed")),
            )

            private fun prim(o: JsonObject, key: String): JsonPrimitive? {
                val el = o[key] ?: return null
                if (el is JsonNull) return null
                return el as? JsonPrimitive
            }

            private fun firstString(o: JsonObject, keys: List<String>): String? {
                for (k in keys) {
                    val p = prim(o, k) ?: continue
                    if (p.isString) {
                        if (p.content.isNotEmpty()) return p.content
                    } else {
                        p.intOrNull?.let { return it.toString() }
                        p.doubleOrNull?.let { return it.toString() }
                    }
                }
                return null
            }

            private fun firstInt(o: JsonObject, keys: List<String>): Int? {
                for (k in keys) {
                    val p = prim(o, k) ?: continue
                    if (p.isString) {
                        p.content.toIntOrNull()?.let { if (it != 0) return it }
                    } else {
                        p.intOrNull?.let { if (it != 0) return it }
                        p.doubleOrNull?.let { val i = it.toInt(); if (i != 0) return i }
                    }
                }
                return null
            }

            private fun firstDouble(o: JsonObject, keys: List<String>): Double? {
                for (k in keys) {
                    val p = prim(o, k) ?: continue
                    if (p.isString) {
                        p.content.toDoubleOrNull()?.let { return it }
                    } else {
                        p.doubleOrNull?.let { return it }
                        p.intOrNull?.let { return it.toDouble() }
                    }
                }
                return null
            }

            private fun firstBool(o: JsonObject, keys: List<String>): Boolean? {
                for (k in keys) {
                    val p = prim(o, k) ?: continue
                    if (p.isString) {
                        val s = p.content
                        return s == "true" || s == "t" || s == "1"
                    }
                    p.booleanOrNull?.let { return it }
                    p.intOrNull?.let { return it != 0 }
                }
                return null
            }
        }
    }
}
